### PR TITLE
feat(cdk/drag-drop): custom drag clone target

### LINKED ISSUE #20536: feat(cdkDrag): Specify drag clone location
#### Feature Description

When dragging an element using cdkDrag, a clone is appended to body [see here](https://github.com/angular/components/blob/f951bf405ca0da9047afe558aaf2c74e5139b87a/src/cdk/drag-drop/drag-ref.ts#L734) 

Whilst end of body is a sensible default location, it'd be great if this were configurable to a DOM Element or ElementRef.

#### Use Case

We have an Angular component library which is built against a "global" CSS package.

Each selector in the global CSS package is selector scoped against a data attribute describing which brand is belongs to, such as `[data-theme="blue"] .button`

This enables a single microfrontend to be easily theme switched based on user / channel context.
Our `<body>` looks something akin to:
```html
...
<body>

    <div data-theme="blue">
        <app-one-root>
            <button class="button">Button</button>
        </app-one-root>
    <div>

    <div data-theme="red">
        <app-two-root>
            <button class="button">Button</button>
        </app-two-root>
    <div>

</body>
...
```
When using the Drag and Drop module, the clone is created outside either of the scoped themes so therefore the clone is unstyled so you get something like this:

```html
...
<body>

    <div data-theme="blue">
        <app-one-root>
            <button class="button">Button</button>
        </app-one-root>
    <div>

    <div data-theme="red">
        <app-two-root>
            <button class="button">Button</button>
        </app-two-root>
    <div>

    <button class="button cdk-drag cdk-drag-dragging">Button</button> <--- Outside of CSS scope

</body>
...
```

Whilst this problem doesn't happen when using style encapsulated components, we're unable to achieve the desired multi-theming thanks to having to support IE11 and it's lack of support for custom properties.

The only solution I can think of is similar to the CDK Portal, being able to control where it embeds.

### PR DESCRIPTION
**Pull request by geometricpanda**
By default the drag directive clones the element it's applied to and appends to
the end of body. This change allows for a selector, HTML element, or an ElementRef
to be provided for the clone to be appended to instead.

Closes #20536 #13288

### CODE REVIEW

File: src/components-examples/cdk/drag-drop/cdk-drag-drop-custom-clone-target/cdk-drag-drop-custom-clone-target-example.html
Comment (crisbeto) [inline src/components-examples/cdk/drag-drop/cdk-drag-drop-custom-clone-target/cdk-drag-drop-custom-clone-target-example.html:7]: I'm not sure whether this example would illustrate the difference very well. Maybe there should be some inherited styles so it's more visible?

File: src/cdk/drag-drop/directives/drag.ts
Comment (crisbeto) [inline src/cdk/drag-drop/directives/drag.ts:146]: We may also want to add a disclaimer to the docs that if people do this, we can't guarantee that `z-index` and `overflow: hidden` will be accounted for correctly.

File: src/cdk/drag-drop/directives/drag.ts
Comment (geometricpanda) [reply src/cdk/drag-drop/directives/drag.ts:145]: Oh I really like that!
I'll look into it over the weekend

File: src/cdk/drag-drop/directives/drag.ts
Comment (geometricpanda) [reply src/cdk/drag-drop/directives/drag.ts:316]: Sure thing, I'll get this removed. 

I'd put it in the ngOnChanges incase someone updated it post-initialisation, more than happy to update though.

File: src/components-examples/cdk/drag-drop/cdk-drag-drop-custom-clone-target/cdk-drag-drop-custom-clone-target-example.html
Comment (geometricpanda) [reply src/components-examples/cdk/drag-drop/cdk-drag-drop-custom-clone-target/cdk-drag-drop-custom-clone-target-example.html:7]: I actually thought of doing that then thought maybe it was overkill. Sure thing!

File: src/cdk/drag-drop/drag-drop.md
Comment (geometricpanda) [reply src/cdk/drag-drop/drag-drop.md:216]: I noticed that on another file - I've got an aggressive IDE - I'll get this fixed

File: src/cdk/drag-drop/directives/drag.ts
Comment (geometricpanda) [reply src/cdk/drag-drop/directives/drag.ts:146]: Am I best putting that in this comment or in the .md file?

Comment (crisbeto) [reply src/cdk/drag-drop/directives/drag.ts:146]: I think it's better in the comment since it's right next to the code and will be easier to remember to update it if we decide to change the type.

File: src/cdk/drag-drop/directives/drag.ts
Comment (geometricpanda) [reply src/cdk/drag-drop/directives/drag.ts:145]: `global`, `ElementRef` and `HTMLElement` are not a problem, however having it alongside it's self is causing major perf problems on drag.

I *think* this is caused by the inline style `transform` triggering "changes" in the `DragList`. As such i've appended it to the parent of the `DragList`, hope this is ok?

File: src/components-examples/cdk/drag-drop/cdk-drag-drop-custom-clone-target/cdk-drag-drop-custom-clone-target-example.html
Comment (geometricpanda) [reply src/components-examples/cdk/drag-drop/cdk-drag-drop-custom-clone-target/cdk-drag-drop-custom-clone-target-example.html]: this is done now :-)

File: src/cdk/drag-drop/drag-drop.md
Comment (geometricpanda) [inline src/cdk/drag-drop/drag-drop.md:1]: @crisbeto not sure why GitHub is picking up this diff incorrectly now, I've checked locally and the only changes are the additional API docs.
Have you seen this on other PRs?

Comment (geometricpanda) [reply src/cdk/drag-drop/drag-drop.md]: nvm, sorted it using vim.
Webstorm issue

File: src/cdk/drag-drop/directives/drag.ts
Comment (crisbeto) [inline src/cdk/drag-drop/directives/drag.ts:154]: The spacing between elements here is inconsistent.

File: src/cdk/drag-drop/drag-drop.md
Comment (crisbeto) [inline src/cdk/drag-drop/drag-drop.md:133]: We should have another sentence here explaining that the element is appended to the body in order to avoid issues with `z-index` and `overflow: hidden`.

File: src/cdk/drag-drop/drag-drop.md
Comment (crisbeto) [inline src/cdk/drag-drop/drag-drop.md:135]: attribute -> input

File: src/cdk/drag-drop/drag-drop.md
Comment (crisbeto) [inline src/cdk/drag-drop/drag-drop.md:138]: I don't think a `TemplateRef` makes sense since templates aren't inside the DOM.

File: src/cdk/drag-drop/drag-ref.ts
Comment (crisbeto) [inline src/cdk/drag-drop/drag-ref.ts:738]: I don't think the logic for failling back to the configured preview container belongs in `getPreviewInsertionPoint`. It would be cleaner by:

1. Renaming `getPreviewInsertionPoint` to something like `getGlobalPreviewInsertionPoint`.
2. Changing the logic here to:

```
const previewContaier = this.previewContainer || getGlobalPreviewInsertionPoint();
```

File: src/cdk/drag-drop/drag-ref.ts
Comment (crisbeto) [inline src/cdk/drag-drop/drag-ref.ts:301]: This should follow the same pattern as the root and boundary elements where we have a method that accepts both `ElementRef` and `HTMLElement`.

File: src/cdk/drag-drop/directives/drag.ts
Comment (crisbeto) [inline src/cdk/drag-drop/directives/drag.ts:386]: You should be able to clean it up like this:

```
private _getPreviewContainer() {
  const cloneTarget = this.previewContainer;

  if (!cloneTarget || cloneTarget === 'global') {
    return null;
  }

  if (cloneTarget === 'parent') {
    return this.element.nativeElement.parentElement;
  }

  return coerceElement(cloneTarget);
}
```

File: src/cdk/drag-drop/directives/drag.ts
Comment (crisbeto) [inline src/cdk/drag-drop/directives/drag.ts:401]: I think that this should use `this.element.nativeElement.parentElement` since the drag item won't necessarily be directly inside the drop list.

File: src/cdk/drag-drop/directives/drag.ts
Comment (crisbeto) [reply src/cdk/drag-drop/directives/drag.ts]: What sorts of performance issues is it causing? I wouldn't expect an extra style mutation to have much of an effect since we aren't using any mutation observers and the DOM is only checked once dragging has been initiated.

I think that `parent` should mean "next to the current `cdkDrag`" since the `cdkDropList` could have some styles that are inherited to the child elements.

File: src/cdk/drag-drop/directives/drag.ts
Comment (geometricpanda) [reply src/cdk/drag-drop/directives/drag.ts:154]: ooh good catch :-)

File: src/cdk/drag-drop/drag-drop.md
Comment (geometricpanda) [reply src/cdk/drag-drop/drag-drop.md:133]: how's this sound?
```
/**
   *  Target for Drag Preview clone to append into.
   *  `global` will place the clone at the end of `<body>`.
   *  `parent` will place the clone in the parent to the `CdkDropList` .
   *  `ElementRef` or `HTMLElement` will place the clone inside the element.
   *
   *  When `global` (default), the Drag Preview is appended to `<body>` to avoid
   *  potential issues with `z-index` and `overflow: hidden`. By changing the
   *  location you run risk of ancestors styled with `overflow: hidden` hiding part
   *  or all of the Drag Preview, as well as other items styled with a higher
   *  `z-index` overlapping the preview container.
   */
```

File: src/cdk/drag-drop/drag-drop.md
Comment (geometricpanda) [reply src/cdk/drag-drop/drag-drop.md:135]: sorted

File: src/cdk/drag-drop/drag-drop.md
Comment (geometricpanda) [reply src/cdk/drag-drop/drag-drop.md]: Yeah I was using the wrong term, there is not enough coffee in the world!
I've removed `TemplateRef`

File: src/cdk/drag-drop/drag-ref.ts
Comment (geometricpanda) [reply src/cdk/drag-drop/drag-ref.ts]: yeah completely agree - simpler, cleaner 👍

File: src/cdk/drag-drop/drag-ref.ts
Comment (geometricpanda) [reply src/cdk/drag-drop/drag-ref.ts:301]: I was thinking about that, but I was a little apprehensive `ElementRef` has `ElementRef.nativeElement`, and it's already getting coerced to a `HTMLElement` in the `Drag` directive.

I've updated it now, do you think it's worth re-running the `coerceElement` on the `const previewContainer`

File: src/cdk/drag-drop/directives/drag.ts
Comment (geometricpanda) [reply src/cdk/drag-drop/directives/drag.ts:389]: Sure thing,
First PR so I'd assumed based on other parts of the code that the type check was important stylistically.

I've done the stylistic part of this, holding off on the `dropContainer` bit mentioned in another comment until i can get you some performance related stats.

File: src/cdk/drag-drop/directives/drag.ts
Comment (geometricpanda) [reply src/cdk/drag-drop/directives/drag.ts]: Specifically its render delays in the preview element:
- Dragging becomes slow, it takes a while for the drag preview item to catch up with the mouse position
- It appears to be impacting CSS selectors from cascading correctly.

In the demo app there's a selector `.demo-list .cdk-drag-preview` which applies a border when dragging but only when the preview container is set to parent, however when i'm setting the drag container to be the same parent as the `cdkDrag` element, the styles are't getting applied.

When dragging the item out of the list into another and back in, then it's getting applied, i'm assuming that works because of the re-render caused by `transferArrayItem`.

Any advice here would be more than welcomed

File: src/cdk/drag-drop/directives/drag.ts
Comment (geometricpanda) [reply src/cdk/drag-drop/directives/drag.ts:398]: I havent seen any issues when using it outside of a drop list.
It seems that the clone is only created when used with a drop list? I'm still learning the cdk so if i've missed a use case please let me know and i'll check it out

Comment (crisbeto) [reply src/cdk/drag-drop/directives/drag.ts:398]: You're right that the preview is only created when the item is inside of a list. What I meant above is that we'd want the clone to be next to the item being dragged, because the list itself might have some styles that are inherited by the items.

File: src/cdk/drag-drop/drag-ref.ts
Comment (crisbeto) [reply src/cdk/drag-drop/drag-ref.ts:301]: I thought I commented on it already, but I can't find the comment now so here it is again: instead of having the `previewContainer` be a public property, we should use a `withPreviewContainer` method instead, like we're doing for the root element. That would be a good place to run `coerceElement` as well.

File: src/cdk/drag-drop/directives/drag.ts
Comment (crisbeto) [inline src/cdk/drag-drop/directives/drag.ts:157]: nit: the `ElementRef` should be `ElementRef<HTMLElement>`.

File: src/cdk/drag-drop/directives/drag.ts
Comment (crisbeto) [inline src/cdk/drag-drop/directives/drag.ts:274]: You can drop this call which will allow you to remove the `_updatePreviewContainer` method completely. The `_getPreviewContainer` call that you added in the `beforeStarted` subscription should be enough to handle all the cases since it fires right before dragging has started.

File: src/cdk/drag-drop/directives/drag.ts
Comment (geometricpanda) [reply src/cdk/drag-drop/directives/drag.ts:154]: done

File: src/cdk/drag-drop/directives/drag.ts
Comment (geometricpanda) [reply src/cdk/drag-drop/directives/drag.ts:271]: done :-)

File: src/cdk/drag-drop/drag-ref.ts
Comment (geometricpanda) [reply src/cdk/drag-drop/drag-ref.ts:301]: done that now :-)

Comment (geometricpanda) [reply src/cdk/drag-drop/drag-ref.ts:301]: done this - will be in next push

File: src/cdk/drag-drop/directives/drag.ts
Comment (geometricpanda) [reply src/cdk/drag-drop/directives/drag.ts:395]: i guess the outcome of this is related to the other conversation, so i'm gonna mark this as resolved so we're not discussing in two places 👍

File: src/cdk/drag-drop/directives/drag.ts
Comment (geometricpanda) [reply src/cdk/drag-drop/directives/drag.ts]: Hey @crisbeto - I'm about to push all the other changes we've discussed but i still dont have a conclusion on this point, i can make `parent` work as described but it's still a concern that there's serious performance issues when doing it. I've not been able to even diagnose what's causing it.

Presently thinking we could either:
- name it such as `list-sibling` to better describe where it's going
- remove any concep of `parent` and leave it as `global` or a specified location?

Comment (crisbeto) [reply src/cdk/drag-drop/directives/drag.ts]: If you feel that `parent` is problematic, we can leave it out of this PR and add it in a follow-up. That being said, wouldn't it still perform in the same way if you passed in the parent as a raw DOM node?

Comment (geometricpanda) [reply src/cdk/drag-drop/directives/drag.ts]: I'd appreciate that, i guess if we could agree some wording then it might be worth mentioning in the docs.
Something like: "Specifying a target inside the `DragList` can lead to rendering performance issues" ?

**Comment by geometricpanda** [comment]
@crisbeto hows this looking for you?

**Review by crisbeto — COMMENTED** [review]
Looks good, but there are some places that still refer to the `parent` value.

**Comment by geometricpanda** [comment]
> Looks good, but there are some places that still refer to the `parent` value.

Ah, the MD file!
Not sure why but there was a strange caching issue happening on my local machine - probably a dodgy webstorm plugin.

I've fixed that now, the only references to parent left in the code are completely unrelated to the PR.

Thank you for all your help on this btw

**Review by crisbeto — APPROVED** [review]
LGTM. Thank you for the contribution!

**Comment by annieyw** [comment]
@geometricpanda please rebase when possible

**Comment by geometricpanda** [comment]
@annieyw couldnt rebase as i cant get to my dev laptop until after UK lockdown, had to use the github merge tools which created a merge commit, hope that's ok?

**Comment by wagnermaciel** [comment]
@geometricpanda One of your commit messages is not correctly formatted. If possible, please reword the invalid commit message

**Comment by geometricpanda** [comment]
> @geometricpanda One of your commit messages is not correctly formatted. If possible, please reword the invalid commit message

it was having an issue with the github merge commit, now i've finally got my dev laptop back i've been able to rebase so fingers crossed it should be ok now?

**Comment by AlecDusheck** [comment]
Any updates on this PR? It would be nice since I'm having styling issues

**Comment by geometricpanda** [comment]
> Any updates on this PR? It would be nice since I'm having styling issues

There's been some major changes been merged into the way the drag container is being handled in the CDK which means this mechanism is no longer going to work I fear :-(

**Comment by crisbeto** [comment]
@geometricpanda I don't think anything fundamentally changed in the module, but there were definitely some changes that could've caused merge conflicts. Do you intend on reopening this or should I pick up the feature request?

**Comment by crisbeto** [comment]
I ended up re-implementing this myself in #21830. I also added support for `parent` which was omitted here.

**Pull request closed without merging**
